Update JNI target invocation + Add 31-bit caller-specific JNI functions for zOS 31/64 interoperability

joransiu opened the pull request:
This PR introduces base support into VM to support zOS 31/64 interoperability, leveraging LE CEL4RO31 support.

- Introduces a new define `J9VM_ZOS_3164_INTEROPERABILITY` to guard the relevant code paths which are enabled only on zOS 64-bit.
- Adds support for JNI call invocation for JNI native functions, and JNI_*Load routines leveraging the CEL4RO31 support in libffi being introduced under: https://github.com/eclipse/openj9/pull/12025
- Introduces a parallel subset of JNICALL functions that need to be aware the native caller is 31-bit.  Specifically:
   - `call*MethodV` / `newObjectV` to interpret va_args in 32-bit format.  `pushArguments` is updated with additional logic to interpret 32-bit va_args.
   - `get/releaseArrayElements` + `get/releaseStringChars*` to guarantee return storage is within 31-bit address-ability.
- Introduces new fields in J9VMThread and J9JavaVM, along with helper query functions, to track the matching 31-bit counterparts from 31-bit libjvm31 shim library.

Signed-off-by: Joran Siu <joransiu@ca.ibm.com>

joransiu commented:
@pshipton / @gacholio : Do you mind reviewing these changes?  Thanks!

gacholio commented on runtime/vm/BytecodeInterpreter.hpp:
```diff
@@ -2382,9 +2382,23 @@ class INTERPRETER_CLASS
+				if (NULL == _currentThread->jniEnv31) {
+					queryJNIEnv31(_currentThread);
+				}
+				values[0] = &(_currentThread->jniEnv31);
```
Does this not require a `void*` cast?

joransiu replied:
The compiler didn't complain, but I added the cast to `void*` for better consistency.  Fixed under: 48f9cc6c856c83d1079c8f8b9ce6bde733d9855b

gacholio commented on runtime/vm/threadhelp.cpp:
```diff
@@ -419,4 +423,78 @@ failedToSetAttr(IDATA rc)
 
+#if defined(J9VM_ZOS_3164_INTEROPERABILITY)
+void
+queryJNIEnv31(J9VMThread* vmThread)
```
These functions need some kind of error handling, even if it's just an assert. The callers expect them to succeed.

joransiu replied:
I considered ways to pass the success/failure, but not sure if there's a good way to handle the error on from the callers.  In any case, I expect this to succeed once we have the 31-bit shim library in place.  Added `VM_assert_true` to both functions.  Note, I updated the type of both `vmThread->jniEnv31` and `vm->javaVM31` to U_32 to ensure these 32-bit values are passed properly to the 31-bit targets (via ffi).   Hence, I updated the comparisons to be `0 !=` as opposed to `NULL !=`.  Fixed under: 1220675750c56aa141a7a8819cb60884da621a3f

pshipton commented on runtime/vm/jnicgen.c:
```diff
@@ -138,7 +179,7 @@ jdouble JNICALL callVirtualDoubleMethod (JNIEnv *env, jobject receiver, jmethodI
 	va_list va;
 	va_start(va, methodID);
-	RUN_CALLIN_METHOD(env, receiver, NULL, methodID, VA_PTR(va));
+	RUN_CALLIN_METHOD(env, receiver, NULL, methodID, (U_8*)(VA_PTR(va)) + 2);
```
Is this supposed to be modified?

joransiu replied:
Fixed under: e4ba6504f90ab6bedab07f23600fef8cc59548fa  Thanks!

pshipton commented on runtime/vm/jnimisc.cpp:
```diff
@@ -636,20 +636,24 @@ setArrayRegion(JNIEnv *env, jarray array, jsize start, jsize len, void *buf)
-getArrayElements(JNIEnv *env, jarray array, jboolean *isCopy)
-{
+static void*
+getArrayElementsCore(JNIEnv *env, jarray array, jboolean *isCopy, jboolean ensureMem32) {
```
{ on next line

joransiu replied:
Fixed under c55d898c261fd84045221f47f703e48ef742eadd

joransiu commented:
Thanks for your reviews @pshipton & @gacholio.   I've updated rebased the code against the lastest:
- Updated the types of `jniEnv31` and `javaVM31` handles to U_32, as they should be 32-bit pointers.
- Fixed a missing `result = 0` for JNI_OnUnLoad calls, and passing of NULL parameter to the ffi values within `sendLifecycleEventCallback`.

gacholio commented:
I'll do another review pass today. Tagging @fjeremic in case he has comments.

gacholio commented on runtime/oti/j9nonbuilder.h:
```diff
@@ -4923,6 +4923,9 @@ typedef struct J9VMThread {
 #endif /* !J9VM_ENV_DATA64 */
 #endif /* J9VM_JIT_RUNTIME_INSTRUMENTATION */
+#if defined(J9VM_ZOS_3164_INTEROPERABILITY)
+	U_32 jniEnv31;
```
This will misalign the rest of the structure.

joransiu replied:
Added U_32 pad member immediately after.  Since `J9VM_ZOS_3164_INTEROPERABILITY` applies to 64-bit, I didn't see how to optimize it with the prior paddings via ifdef's, but left a comment.  https://github.com/eclipse/openj9/pull/12074/commits/1bb3113393e6c199034b4b023b886b6a656b2ee7

gacholio commented on runtime/oti/j9nonbuilder.h:
```diff
@@ -5443,6 +5446,9 @@ typedef struct J9JavaVM {
 	UDATA vmtargetOffset;
 #endif /* defined(J9VM_OPT_OPENJDK_METHODHANDLE) */
+#if defined(J9VM_ZOS_3164_INTEROPERABILITY)
+	U_32 javaVM31;
```
This will misalign the rest of the structure (once more is added).

joransiu replied:
Added U_32 pad member immediately after.  https://github.com/eclipse/openj9/pull/12074/commits/1bb3113393e6c199034b4b023b886b6a656b2ee7

gacholio commented on runtime/vm/threadhelp.cpp:
```diff
@@ -419,4 +423,80 @@ failedToSetAttr(IDATA rc)
+	UDATA getJavaVM31Handle = 0;
+
+	if (0 == vm->javaVM31) {
+
```
Please remove this blank line.

joransiu replied:
Updated in https://github.com/eclipse/openj9/pull/12074/commits/e2c7aa733250dfc7af3d882992feb4d5195e5fda

gacholio commented on runtime/vm/threadhelp.cpp:
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 1998, 2020 IBM Corp. and others
+ * Copyright (c) 1998, 2021 IBM Corp. and others
```
Why is this file being changed? Wouldn't this be better in one of the JNI support files?

joransiu replied:
I was thinking it tied to J9VMThread at first... would `jnimisc.cpp` be a better home for these functions?

gacholio replied:
I think so, yes.

joransiu replied:
I migrated over the implementations to `jnimisc.cpp`, and the function prototypes to `vm_internal.h`.   Considered making them VM internal functions, as the JIT DirectToJNI dispatch may need to query the matching 31-bit jniEnv if not resolved at JNI call.  Although, given the complexity of CEL4RO31 dispatch and handling, will need to consider which scenarios we want / can effectively build JIT dispatch support.  I propose we move only expose these functions to VM internal functions if needed.

https://github.com/eclipse/openj9/pull/12074/commits/e2c7aa733250dfc7af3d882992feb4d5195e5fda

joransiu commented on runtime/vm/jnimisc.cpp:
```diff
@@ -636,20 +636,25 @@ setArrayRegion(JNIEnv *env, jarray array, jsize start, jsize len, void *buf)
-void* JNICALL
-getArrayElements(JNIEnv *env, jarray array, jboolean *isCopy)
+static void*
+getArrayElementsCore(JNIEnv *env, jarray array, jboolean *isCopy, jboolean ensureMem32)
```
Makes sense, I'll revise.  Thanks.

joransiu replied:
Updated in https://github.com/joransiu/openj9/commit/5300ef9f6450d256f95d497c001311ad3dc7c89d

joransiu commented on runtime/vm/vmbootlib.c:
```diff
@@ -144,14 +148,49 @@ sendLifecycleEventCallback(struct J9VMThread* vmThread, struct J9NativeLibrary*
-		} else {
-			result = JNI_Load((JavaVM*)vmThread->javaVM, NULL);
+#if defined(J9VM_ZOS_3164_INTEROPERABILITY)
+		if (J9_ARE_ALL_BITS_SET((UDATA)JNI_Load, J9PORT_SL_ZOS_31BIT_TARGET_HIGHTAG)) {
```
Updated under https://github.com/eclipse/openj9/pull/12074/commits/b511692fed5723e6cb73ee541d35b15e0e1f2f42

joransiu commented on runtime/vm/threadhelp.cpp:
```diff
@@ -419,4 +423,80 @@ failedToSetAttr(IDATA rc)
+			}
+		}
+	}
+	Assert_VM_true(0 != vm->javaVM31);
```
Updated in https://github.com/eclipse/openj9/pull/12074/commits/e2c7aa733250dfc7af3d882992feb4d5195e5fda

joransiu commented on runtime/vm/threadhelp.cpp:
```diff
@@ -419,4 +423,80 @@ failedToSetAttr(IDATA rc)
+			}
+		}
+	}
+	Assert_VM_true(0 != vmThread->jniEnv31);
```
Updated in https://github.com/eclipse/openj9/pull/12074/commits/e2c7aa733250dfc7af3d882992feb4d5195e5fda

joransiu commented on runtime/vm/threadhelp.cpp:
```diff
@@ -419,4 +423,80 @@ failedToSetAttr(IDATA rc)
+	UDATA getJNIEnv31Handle = 0;
+
+	if (0 == vmThread->jniEnv31) {
+
```
Updated in https://github.com/eclipse/openj9/pull/12074/commits/e2c7aa733250dfc7af3d882992feb4d5195e5fda

joransiu commented on runtime/vm/jnicsup.cpp:
```diff
@@ -236,6 +236,22 @@ static jobject JNICALL newObjectV(JNIEnv *env, jclass clazz, jmethodID methodID,
+	if (obj) {
+		CALL_NONVIRTUAL_VOID_METHOD_V31(env, obj, METHODID_CLASS_REF(methodID), methodID, args);
+		if (exceptionCheck(env)) {
+				deleteLocalRef(env, obj);
```
Fixed in https://github.com/eclipse/openj9/pull/12074/commits/699fcb6ced68c473ae0f1017e716bc22dfba78b0

joransiu commented:
Updated the review comments, and fixed a couple additional issues:
- JNI_*Load ffi call had incorrect number of arguments.
- Cleaned up ffi prep to not instantiate arg/values if not needed.
- Added 31-bit dispatch logic for JNI exit / abort hooks, for 31-bit functions registered during JNI_CreateJavaVM.   https://github.com/eclipse/openj9/pull/12074/commits/34d2a60f4165420039122f4cf02ff9e4ae7a16b1

One outstanding issue:
I currently have `jniArrayAllocateMemory32FromThread` and `jniArrayFreeMemory32FromThread` prototypes defined under `vm_api.h` (https://github.com/eclipse/openj9/pull/12074/commits/3afd2e3043e93e42a5025885ef87ed6a6b4f7c8c), though I haven't yet added the corresponding entries in VM internal functions -- to match their "non-32" counterparts.   I don't expect GC will need to call these routines given we cannot be able to hand back critical regions to 31-bit targets anyway.  The only callers so far are all within `vm`.  As such, we could consider just defining them in `vm_internal.h` instead then.  @gacholio, any preference on this?

gacholio commented:
Let's put them in the internal header.

joransiu commented:
@gacholio : I've re-based the changes against the latest master, and added two additional commits on top.  The first is to introduce a `-XX:+Enable3164Interoperability` option (disabled by default) to attempt loading 31-bit native libraries and also disable DirectToJNI support (as JIT support is not available yet).  The second commit creates a customized below-the-bar JNIInterface_ function table that the 31-bit shim library will use.   Do you mind reviewing the latest updates?  Thanks!

pshipton commented on runtime/vm/jnicsup.cpp:
```diff
@@ -165,6 +165,31 @@ static jobject JNICALL getModule(JNIEnv *env, jclass clazz);
+#define RELEASE_LONG_ARRAY_ELEMENTS31 ((void (JNICALL *)(JNIEnv *env, jlongArray array, jlong * elems, jint mode))releaseArrayElements31)
+#define RELEASE_FLOAT_ARRAY_ELEMENTS31 ((void (JNICALL *)(JNIEnv *env, jfloatArray array, jfloat * elems, jint mode))releaseArrayElements31)
+#define RELEASE_DOUBLE_ARRAY_ELEMENTS31 ((void (JNICALL *)(JNIEnv *env, jdoubleArray array, jdouble * elems, jint mode))releaseArrayElements31)
+#define RELEASE_PRIMITIVE_ARRAY_CRITICAL ((void (JNICALL *)(JNIEnv *env, jarray carray, void * elems, jint mode))releaseArrayElements31)
```
Shouldn't this and the next two macros contain 31 in the name? It would be less confusing.

joransiu replied:
Updated in 8c44f2d.

pshipton commented on runtime/oti/j9port.h:
```diff
@@ -759,6 +759,12 @@ typedef struct J9CacheInfoQuery {
 #define J9PORT_SLOPEN_OPEN_EXECUTABLE OMRPORT_SLOPEN_OPEN_EXECUTABLE
 
+#if defined(J9ZOS39064)
+#define J9PORT_SLOPEN_ATTEMPT_31BIT_OPEN OMRPORT_SLOPEN_ATTEMPT_31BIT_OPEN
```
This isn't required, the code can just use the OMR constants directly.

joransiu replied:
Oh, okay.  I was just following the examples above. :)

pshipton replied:
We didn't want to change all the code when OMR was split, but anything new can use OMR directly.

joransiu replied:
I dropped these 3 macros in j9port.h and updated the couple places where I used the newly added J9PORT macros in a16bb7a2b122857ac766df05c80832235b8ea4d3,  dbaa1783c862d2d927895fef6ae6a98939ef8496 and 14598851209ed963aff5ac2c8501b7a662c40332.

pshipton commented:
@gacholio any further concerns? Internal testing passed (with this change and #12554, and internal change 162) in internal build 470470.

gacholio commented:
I'll give this one more look today.

gacholio commented on runtime/oti/j9.h:
```diff
@@ -349,6 +349,10 @@ typedef struct {
 } J9_UNMODIFIABLE_CLASS_ANNOTATION_DATA;
 
+#if defined(J9VM_ZOS_3164_INTEROPERABILITY)
+#define J9_IS_31BIT_INTEROP_TARGET(handle) J9_ARE_ALL_BITS_SET((UDATA)handle, OMRPORT_SL_ZOS_31BIT_TARGET_HIGHTAG)
```
handle needs to be bracketted.

joransiu replied:
Good catch.  Updated in https://github.com/eclipse/openj9/pull/12074/commits/cfd0e84fed6fb508a2bc233bf2a057cffed13603 .  Thanks!

gacholio commented on runtime/vm/jnimem.c:
```diff
@@ -51,6 +51,10 @@ void* jniArrayAllocateMemoryFromThread(J9VMThread* vmThread, UDATA sizeInBytes)
 #endif
 }
 
+void* jniArrayAllocateMemory32FromThread(J9VMThread* vmThread, UDATA sizeInBytes) {
```
Should these functions (and their prototypes) be ifdeffed for consistency?

joransiu replied:
@gacholio: The code consuming these two Memory32 APIs are in the refactored get/releaseArrayElements/StringImpl, that takes in the newly added `ensureMem32` parameter.  Arguably, these two are generic enough utility functions -- and if I #ifdeffed them out, the base implementation won't actually guarantee correctness if a caller passes true to `ensureMem32`.  Thoughts?

```C
+static void*
+getArrayElementsImpl(JNIEnv *env, jarray array, jboolean *isCopy, jboolean ensureMem32)
 {
....
              if (ensureMem32) {
                       elems = jniArrayAllocateMemory32FromThread(currentThread, ROUND_UP_TO_POWEROF2(byteCount, sizeof(UDATA)));
               } else {
                       elems = jniArrayAllocateMemoryFromThread(currentThread, ROUND_UP_TO_POWEROF2(byteCount, sizeof(UDATA)));
               }
```

gacholio replied:
Fair enough, I suppose - leave them alone.

gacholio commented:
jenkins compile win jdk8

gacholio commented:
jenkins compile win jdk8
